### DISCUSSION TITLE
Lookup for URL by router and endpoint name

### DISCUSSION THREAD

**Discussion by ganiular** [fastapi/fastapi#9489] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter, Request
from fastapi.responses import RedirectResponse

...
```

### Description

I wish the APIRouter can have a name property, which can be use for referencing just as the flash Blueprint(name="product") which then can be use in url_for("product.endpoint_name").

What I really want to achieve is this
In products.py

    router = APIRouter(name="product", prefix='/product')

    @router.get('/store/{store_id}')
    async def list_store_products(store_id: str):
        return [...]
Now in stores.py

    router = APIRouter(name="store", prefix='/store')

    @router.get('/{store_id}/products')
    async def list_store_products(store_id: str, req: Request):
        return RedirectResponse(req.url_for("product.list_store_products"))

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

3.10

**Comment by abhint** [comment]
Currently, FastAPI's APIRouter does not have a built-in name property like Flask's Blueprint. However, you can achieve similar functionality by using FastAPI's dependency injection system and referencing the endpoint path directly.
```py
from fastapi import APIRouter, Request, FastAPI
from fastapi.responses import RedirectResponse

# Create a custom APIRouter subclass
class NamedAPIRouter(APIRouter):
    def __init__(self, name: str = None, **kwargs):
        super().__init__(**kwargs)
        self.name = name

# Create your routers using the custom NamedAPIRouter
product_router = NamedAPIRouter(name="product", prefix='/product')
store_router = NamedAPIRouter(name="store", prefix='/store')

@product_router.get('/store/{store_id}')
async def list_store_products(store_id: str):
    return [...]

@store_router.get('/{store_id}/products')
async def list_store_products(store_id: str, request: Request):
    # Access the named router using the name property and generate the URL
    product_url = request.url_for(f"{product_router.name}.list_store_products", store_id=store_id)
    return RedirectResponse(product_url)
```
